azurerm/internal/services/network/data_source_service_tags.go

```diff
@@ -0,0 +1,89 @@
+package network
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/helpers/azure"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/clients"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/timeouts"
+)
+
+func dataSourceServiceTags() *schema.Resource {
+	return &schema.Resource{
+		Read: dataSourceServiceTagsRead,
+
+		Timeouts: &schema.ResourceTimeout{
+			Read: schema.DefaultTimeout(5 * time.Minute),
+		},
+
+		Schema: map[string]*schema.Schema{
+			"location": azure.SchemaLocation(),
+
+			"service": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+
+			"region": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+
+			"address_prefixes": {
+				Type:     schema.TypeList,
+				Computed: true,
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
+			},
+		},
+	}
+}
+
+func dataSourceServiceTagsRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*clients.Client).Network.ServiceTagsClient
+	ctx, cancel := timeouts.ForRead(meta.(*clients.Client).StopContext, d)
+	defer cancel()
+
+	location := d.Get("location").(string)
+	res, err := client.List(ctx, location)
+	if err != nil {
+		return err
+	}
+
+	service := d.Get("service").(string)
+	region := d.Get("region").(string)
+
+	if res.Values == nil {
+		d.Set("address_prefixes", make([]string, 0))
+		return nil
+	}
+
+	addressPrefixes := make([]string, 0)
+	for _, sti := range *res.Values {
+		if props := sti.Properties; props != nil {
+			if *props.SystemService != service {
+				continue
+			}
+
+			if *props.Region == region {
+				addressPrefixes = *props.AddressPrefixes
+			}
+		}
+	}
+
+	err = d.Set("address_prefixes", addressPrefixes)
+	if err != nil {
+		return fmt.Errorf("Error setting `address_prefixes`: %+v", err)
+	}
```

Comment:
Can we just do following change? 

```suggestion
	for _, sti := range *res.Values {
		if sti.Name == nil || !isServiceTagOf(*sti.Name, service) {
			continue
		}

		if props := sti.Properties; props != nil {
			if *props.Region == region {
				addressPrefixes := make([]string, 0)
				if props.AddressPrefixes != nil {
					addressPrefixes = *props.AddressPrefixes
				}
				err = d.Set("address_prefixes", addressPrefixes)
				if err != nil {
					return fmt.Errorf("Error setting `address_prefixes`: %+v", err)
				}
				if sti.ID == nil {
					return errors.New("unexpected nil ID for service tag")
				}
				d.SetId(*sti.ID)
				return nil
			}
		}
	}
	return errors.New("specified service tag not found")
}

// isServiceTagOf is used to check whether a service tag name belongs to the service of name `serviceName`.
// Service tag name has format as below:
// - (regional) serviceName.locationName
// - (all) serviceName
func isServiceTagOf(stName, serviceName string) bool {
	stNameComponents := strings.Split(stName, ".")
	if len(stNameComponents) != 1 && len(stNameComponents) != 2 {
		return false
	}
	return stNameComponents[0] == serviceName
}
```

There are several changes here:

1. Instead of using `SystemService` to compare against `service` name, I prefer using the `Name` of the `ServiceTagInformation`. There are two reasons to do so:

   - This allow user to refer to the list in [doc](https://docs.microsoft.com/en-us/azure/virtual-network/service-tags-overview) for a specific tag name, with no need to prefix of `Azure`.
   - Based on above point, some tag name (e.g. Sql) has different case between it in `SystemService`(AzureSQL) and `Name`(Sql). Since `Name` is always the same case as it is in the doc, hence I prefer that.

2. Stop the loop once we got the tag info.
3. Return error if no service tag is found.